File: backend/app/schemas/marketplace.py
"""
Marketplace Pydantic Schemas

Defines request and response schemas for marketplace API endpoints.
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class ListingType(str, Enum):
    AGENT = "AGENT"
    SUBAGENT = "SUBAGENT"


class Visibility(str, Enum):
    PUBLIC = "PUBLIC"
    PRIVATE = "PRIVATE"
    UNLISTED = "UNLISTED"


class ModerationStatus(str, Enum):
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"


# Marketplace Category Schemas
class MarketplaceCategoryBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=50)
    display_name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    icon: Optional[str] = Field(None, max_length=100)
    is_active: bool = True
    sort_order: int = 0


class MarketplaceCategoryCreate(MarketplaceCategoryBase):
    pass


class MarketplaceCategoryUpdate(BaseModel):
    display_name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    icon: Optional[str] = Field(None, max_length=100)
    is_active: Optional[bool] = None
    sort_order: Optional[int] = None


class MarketplaceCategoryResponse(MarketplaceCategoryBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class MarketplaceCategoryList(BaseModel):
    items: List[MarketplaceCategoryResponse]
    total: int
    page: int
    page_size: int


# Marketplace Tag Schemas
class MarketplaceTagBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=50)
    display_name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    is_active: bool = True


class MarketplaceTagCreate(MarketplaceTagBase):
    pass


class MarketplaceTagUpdate(BaseModel):
    display_name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    is_active: Optional[bool] = None


class MarketplaceTagResponse(MarketplaceTagBase):
    id: int
    usage_count: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class MarketplaceTagList(BaseModel):
    items: List[MarketplaceTagResponse]
    total: int
    page: int
    page_size: int

# Marketplace Listing Schemas
class MarketplaceListingBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    category_id: Optional[int] = None
    tags: Optional[List[str]] = None
    listing_type: ListingType = ListingType.AGENT
    visibility: Visibility = Visibility.PUBLIC
    version: Optional[str] = None
    preview_images: Optional[List[str]] = None
    demo_video_url: Optional[str] = None


class MarketplaceListingCreate(MarketplaceListingBase):
    agent_id: int
    schema_data: Optional[Dict[str, Any]] = None


class MarketplaceListingUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    category_id: Optional[int] = None
    tags: Optional[List[str]] = None
    visibility: Optional[Visibility] = None
    version: Optional[str] = None
    preview_images: Optional[List[str]] = None
    demo_video_url: Optional[str] = None
    schema_data: Optional[Dict[str, Any]] = None


class MarketplaceListingResponse(MarketplaceListingBase):
    id: int
    agent_id: int
    author_id: int
    moderation_status: ModerationStatus
    moderation_notes: Optional[str] = None
    schema_data: Optional[Dict[str, Any]] = None
    install_count: int
    rating_average: float
    rating_count: int
    view_count: int
    published_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class MarketplaceListingList(BaseModel):
    items: List[MarketplaceListingResponse]
    total: int
    page: int
    page_size: int
    has_more: bool


# Marketplace Installation Schemas
class MarketplaceInstallationCreate(BaseModel):
    listing_id: int


class MarketplaceInstallationResponse(BaseModel):
    id: int
    listing_id: int
    user_id: int
    agent_id: int
    installed_at: datetime
    
    class Config:
        from_attributes = True


# Marketplace Review Schemas
class MarketplaceReviewBase(BaseModel):
    rating: int = Field(..., ge=1, le=5)
    review_text: Optional[str] = None


class MarketplaceReviewCreate(MarketplaceReviewBase):
    listing_id: int


class MarketplaceReviewUpdate(BaseModel):
    rating: Optional[int] = Field(None, ge=1, le=5)
    review_text: Optional[str] = None


class MarketplaceReviewResponse(MarketplaceReviewBase):
    id: int
    listing_id: int
    user_id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class MarketplaceReviewList(BaseModel):
    items: List[MarketplaceReviewResponse]
    total: int
    page: int
    page_size: int


# Moderation Schemas
class ModerationAction(BaseModel):
    moderation_status: ModerationStatus
    moderation_notes: Optional[str] = None


# Search and Filter Schemas
class MarketplaceSearchParams(BaseModel):
    query: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = None
    listing_type: Optional[ListingType] = None
    sort_by: Optional[str] = "created_at"  # created_at, rating, installs, views
    sort_order: Optional[str] = "desc"  # asc, desc
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)

File: frontend/src/types/marketplace.ts
/**
 * Marketplace Types
 * Type definitions for marketplace functionality in Chronos AI Agent Builder Studio
 */

// ============== Enum Types ==============

export type ListingType = 'AGENT' | 'SUBAGENT';

export type Visibility = 'PUBLIC' | 'PRIVATE' | 'UNLISTED';

export type ModerationStatus = 'PENDING' | 'APPROVED' | 'REJECTED';

// ============== Marketplace Category Types ==============

export interface MarketplaceCategory {
    id: number;
    name: string;
    display_name: string;
    description?: string;
    icon?: string;
    is_active: boolean;
    sort_order: number;
    created_at: string;
    updated_at: string;
}

export interface MarketplaceCategoryCreate {
    name: string;
    display_name: string;
    description?: string;
    icon?: string;
    is_active?: boolean;
    sort_order?: number;
}

export interface MarketplaceCategoryUpdate {
    display_name?: string;
    description?: string;
    icon?: string;
    is_active?: boolean;
    sort_order?: number;
}

export interface MarketplaceCategoryList {
    items: MarketplaceCategory[];
    total: number;
    page: number;
    page_size: number;
}

// ============== Marketplace Tag Types ==============

export interface MarketplaceTag {
    id: number;
    name: string;
    display_name: string;
    description?: string;
    is_active: boolean;
    usage_count: number;
    created_at: string;
    updated_at: string;
}

export interface MarketplaceTagCreate {
    name: string;
    display_name: string;
    description?: string;
    is_active?: boolean;
}

export interface MarketplaceTagUpdate {
    display_name?: string;
    description?: string;
    is_active?: boolean;
}

export interface MarketplaceTagList {
    items: MarketplaceTag[];
    total: number;
    page: number;
    page_size: number;
}

// ============== Marketplace Listing Types ==============

export interface MarketplaceListing {
    id: number;
    agent_id: number;
    author_id: number;
    title: string;
    description?: string;
    category_id?: number;
    tags?: string[];
    listing_type: ListingType;
    visibility: Visibility;
    version?: string;
    preview_images?: string[];
    demo_video_url?: string;
    schema_data?: Record<string, unknown>;
    moderation_status: ModerationStatus;
    moderation_notes?: string;
    install_count: number;
    rating_average: number;
    rating_count: number;
    view_count: number;
    published_at?: string;
    created_at: string;
    updated_at: string;
}

export interface MarketplaceListingCreate {
    agent_id: number;
    title: string;
    description?: string;
    category_id?: number;
    tags?: string[];
    listing_type?: ListingType;
    visibility?: Visibility;
    version?: string;
    preview_images?: string[];
    demo_video_url?: string;
    schema_data?: Record<string, unknown>;
}

export interface MarketplaceListingUpdate {
    title?: string;
    description?: string;
    category_id?: number;
    tags?: string[];
    visibility?: Visibility;
    version?: string;
    preview_images?: string[];
    demo_video_url?: string;
    schema_data?: Record<string, unknown>;
}

export interface MarketplaceListingList {
    items: MarketplaceListing[];
    total: number;
    page: number;
    page_size: number;
    has_more: boolean;
}

// ============== Marketplace Installation Types ==============

export interface MarketplaceInstallation {
    id: number;
    listing_id: number;
    user_id: number;
    agent_id: number;
    installed_at: string;
}

export interface MarketplaceInstallationCreate {
    listing_id: number;
}

export interface MarketplaceInstallationResponse {
    id: number;
    listing_id: number;
    user_id: number;
    agent_id: number;
    installed_at: string;
}

// ============== Marketplace Review Types ==============

export interface MarketplaceReview {
    id: number;
    listing_id: number;
    user_id: number;
    rating: number;
    review_text?: string;
    created_at: string;
    updated_at: string;
}

export interface MarketplaceReviewCreate {
    listing_id: number;
    rating: number;
    review_text?: string;
}

export interface MarketplaceReviewUpdate {
    rating?: number;
    review_text?: string;
}

export interface MarketplaceReviewList {
    items: MarketplaceReview[];
    total: number;
    page: number;
    page_size: number;
}

// ============== Search and Filter Types ==============

export interface MarketplaceSearchParams {
    query?: string;
    category?: string;
    tags?: string[];
    listing_type?: ListingType;
    sort_by?: string; // created_at, rating, installs, views
    sort_order?: string; // asc, desc
    page?: number;
    page_size?: number;
}

// ============== Agent Types ==============

export interface Agent {
    id: number;
    name: string;
    description?: string;
    status: string;
    model_config?: Record<string, unknown>;
    system_prompt?: string;
    user_prompt_template?: string;
    sub_agent_config?: Record<string, unknown>;
    tags?: string[];
    metadata?: Record<string, unknown>;
    usage_count: number;
    success_rate: number;
    avg_response_time: number;
    owner_id: number;
    created_at: string;
    updated_at: string;
    version?: string;
    icon?: string;
    color?: string;
    preview_image?: string;
}

// ============== Copy Statistics Types ==============

export interface CopyStatistics {
    listing_id: number;
    total_installs: number;
    recent_installs: number;
    unique_users: number;
    listing_title: string;
}

// ============== Copied Agents List Types ==============

export interface CopiedAgent {
    agent_id: number;
    name: string;
    description: string;
    status: string;
    created_at: string;
    updated_at: string;
    listing_id: number;
    listing_title: string;
    original_agent_id: number;
    installed_at: string;
    author_id: number;
    is_copy: boolean;
    metadata: Record<string, any>;
}

export interface CopiedAgentsList {
    items: CopiedAgent[];
    total: number;
    page: number;
    page_size: number;
    has_more: boolean;
}

// ============== UI State Types ==============

export interface PublishingFormState {
    agentId: number;
    title: string;
    description: string;
    categoryId?: number;
    selectedTags: string[];
    listingType: ListingType;
    visibility: Visibility;
    version: string;
    previewImages: string[];
    demoVideoUrl?: string;
    pricingModel: 'free' | 'paid';
    price?: number;
    licenseType: string;
    supportInformation: string;
    isSubmitting: boolean;
    errors: Record<string, string>;
}

export interface PublishModalState {
    isOpen: boolean;
    publishToChannels: boolean;
    publishToMarketplace: boolean;
    formState: PublishingFormState;
    availableCategories: MarketplaceCategory[];
    availableTags: MarketplaceTag[];
    isLoadingCategories: boolean;
    isLoadingTags: boolean;
}